### ISSUE TITLE
Support for "signature_algorithms_cert" extension (TLS 1.3)

### ISSUE THREAD

**Issue by koes-soptim** [bcgit/bc-java#1456]
Hi!

We are currently developing a software which uses BC to establish TLS/mTLS connections to our partners. In the implementation, we have the strict requirement to always add the "signature_algorithms_cert" extension (IANA #50) to indicate the supported signature algorithms for certificate verification when using TLS 1.3. The extension has to be added even if the value does not differ from the value set in the "signature_algorithms". 

According to our understanding of the code, it is not yet possible to add the extension to the relevant messages (CertificateRequest). I found source code (e.g. _ProvTlsServer.getCertificateRequest()_ for the server, _ProvTlsClient.getSupportedSignatureAlgorithmsCert()_ for the client), where the related values are always _null_.

Therefore I want to ask the following:

1. If my observations are correct, can you add this feature? (If possible and suitable, configurable via a property)

2. Is there anything we can do as a workaround? (e.g. via extending specific classes / overriding specific functions)

Kind regards,

Robin

**Comment by peterdettman** [comment]
If you are using the BCJSSE provider (via the JSSE API) the issue is that JSSE doesn't (AFAIK) provide a way to configure the signature_algorithms_cert extension, but we have the option of supporting it with a BC-specific extension (e.g. a BCSSLParameters signatureSchemesCert property like the SSLParameters signatureSchemes property, and/or a "org.bouncycastle.jsse.client.SignatureSchemesCert" system property like "jdk.tls.client.SignatureSchemes" - and "server" versions).

**Comment by koes-soptim** [comment]
Thank you for the quick reply! @peterdettman 

1. The proposed solution (BCSSLParameters property and/or the system property) would be very helpful. Are there yet any considerations to implement this (e.g. with the next release)?

2. Could it be also another approach to use a JDK-like solution? The _sun.security.ssl.CertSignAlgsExtension_ class seems to just set the same values as in the signature_algorithms extension.

**Comment by peterdettman** [comment]
Since RFC 8446 4.2.3. says:
Implementations which have the same policy in both cases MAY omit the "signature_algorithms_cert" extension.

I guess we could add system properties to configure that choice.

**Comment by peterdettman** [comment]
I've added the following system properties for omitting signature_algorithms_cert:
- org.bouncycastle.jsse.client.omitSigAlgsCertExtension (boolean default true)
- org.bouncycastle.jsse.server.omitSigAlgsCertExtension (boolean default true)

So you would set the relevant property to false if you wanted signature_algorithms_cert redundantly included in the extensions.

Unfortunately we just released 1.76 so a new full release will likely be some time off. We'll see about a beta release maybe at the end of the week to let you test out the changes at least.
